OAuthClient now cache header aware.

kingosticks opened the pull request:
This branch is still a work in progress but any thoughts are very welcome.

jodal commented on mopidy_spotify/web.py:
```diff
@@ -9,11 +9,15 @@
 from mopidy_spotify import utils
 
 logger = logging.getLogger(__name__)
+INFO = logging.getLevelName('INFO')
```
This is already available as `logging.INFO`.

kingosticks replied:
Thanks!

earthican commented:
I recently have just come across the playlists issue, so thank you so much @kingosticks for working on this.

Do you have a timeline on when this will get merged and released?

kingosticks commented:
Sorry, I don't have a timeline for you. I've got plenty on this week, I'll spend time when I can.

earthican commented:
No worries, thanks for the update, and I appreciate your work on this!!

kingosticks commented:
So I (force, sorry) pushed some more fixes and cleanups. It's performing nicely now, the last commit made a big difference to this (for certain clients). Functionally I think I'm done, just the small issue of tests remain...

brownjd commented:
Tried the following using this branch:

`
core = self.frontend.core
core.tracklist.clear()
core.tracklist.add(uri=self.playlists[self.selected].uri)
core.playback.play()
`

where the url was of the form 'spotify:user:username:playlist:xxxxxxxxxx'

Here's the error:

> ERROR
    SpotifyBackend backend caused an exception.
Traceback (most recent call last):
  File "/home/pi/.local/lib/python2.7/site-packages/mopidy/core/library.py", line 19, in _backend_error_handling
    yield
  File "/home/pi/.local/lib/python2.7/site-packages/mopidy/core/library.py", line 236, in lookup
    result = future.get()
  File "/home/pi/.local/lib/python2.7/site-packages/pykka/threading.py", line 52, in get
    compat.reraise(*self._data['exc_info'])
  File "/home/pi/.local/lib/python2.7/site-packages/pykka/compat.py", line 12, in reraise
    exec('raise tp, value, tb')
  File "/home/pi/.local/lib/python2.7/site-packages/pykka/actor.py", line 201, in _actor_loop
    response = self._handle_receive(message)
  File "/home/pi/.local/lib/python2.7/site-packages/pykka/actor.py", line 295, in _handle_receive
    return callee(*message['args'], **message['kwargs'])
  File "/usr/local/lib/python2.7/dist-packages/Mopidy_Spotify-3.1.0-py2.7.egg/mopidy_spotify/library.py", line 34, in lookup
    uri)
  File "/usr/local/lib/python2.7/dist-packages/Mopidy_Spotify-3.1.0-py2.7.egg/mopidy_spotify/lookup.py", line 34, in lookup
    return _lookup_playlist(session, web_session, config, uri)
NameError: global name '_lookup_playlist' is not defined

Is there a better way to play a playlist?

brownjd commented:
Figured it out:

def lookup_playlist(session, web_session, config, uri):
needs to be
def _lookup_playlist(session, web_session, config, uri):

line 94 in lookup.py

brownjd commented:
This is working well. Any chance of supporting Spotify Radio? The last issue was closed a few years ago because the underlying library did not support it. It seems like you're working around that now...

justincrosby commented:
@brownjd's fix to this branch works. Playlists loaded via mopidy/home-assistant are working again. Thanks @kingosticks

kingosticks commented:
I've split the web client caching parts of this into https://github.com/mopidy/mopidy-spotify/pull/195 and added tests.

The playlist part will follow.

kingosticks commented:
Does anyone have any thought about the `WebSession` thing I've created and passed around rather than doing it more like images.py, which is the other option?

kingosticks commented:
@jodal @adamcik

princemaxwell commented:
@kingosticks 
i am using mopidy, mopidy-spotify and their dependencies for playing music with MPC, the command line from MPD.

when i try to add a playlist by "sudo mpc -h 10.0.0.111 -p 6601 add "spotify:playlist:37i9dQZF1DX0XUsuxWHRQd" to MPD i only get:

 "error adding spotify:playlist:37i9dQZF1DX0XUsuxWHRQd: directory or file not found".

Is there a fix for that, too?

kingosticks commented:
@princemaxwell The old style i.e. spotify:user:spotify:playlist:37i9dQZF1DX0XUsuxWHRQd will work. But that form needs adding too. Thanks

princemaxwell commented:
Sure, if links were broken too.

Do this commit really effect MPC????
Because i removed Mopidy-Spotify (only this package). And MPC was still working for artist, track and album URIs of Spotify...

kingosticks commented:
I am afraid I don't understand what you are asking. However, that last commit should fix your other case now too.

princemaxwell commented:
> 
> 
> I am afraid I don't understand what you are asking. However, that last commit should fix your other case now too.

I installed your fix, but it doesn't work for me.
Let me explain, what i try to do.

I have installed a mopidy, mopidy-spotify, all dependencies and WEB ui "Iris". This all together works fine at all.

What i want to do is starting spotify with MPC command line.

```
sudo systemctl restart mopidy
pi@box:~ $ sudo mpc -h box -p 6601 add "spotify:artist:0oLWXBuoO0namCRC2oji0q"
pi@box:~ $ sudo mpc -h box -p 6601 clear
volume:100%   repeat: off   random: off   single: off   consume: off
pi@box:~ $ sudo mpc -h box -p 6601 add "spotify:playlist:37i9dQZF1DX0XUsuxWHRQd"
error adding spotify:playlist:37i9dQZF1DX0XUsuxWHRQd: directory or file not found
pi@box:~ $ sudo mpc -h box -p 6601 add "spotify:user:spotify:playlist:37i9dQZF1DX0XUsuxWHRQd"
error adding spotify:user:spotify:playlist:37i9dQZF1DX0XUsuxWHRQd: directory or file not found
```

you see, adding artist works fine. albums and tracks too.
but when i try to add a playlist with MPC, i get an error.

for own testing, i removed mopidy-spotify (not his dependencies) from my box. i restarted mopidy. the functionality of MPC was the same like before, everything ok but playlists don't work.

so the question is, does MPC use mopidy-spotify für interpeting Spotify URIs?? or does MPC use pyspotify2.0.5 and libspotify12 for this???

kingosticks commented:
And to answer your other question, mpc talks to mopidy which uses
mopidy-spotify for everything Spotify related. If you 'removed'
mopidy-spotify and mpc's behaviour was unchanged then you did not actually
remove the version of mopidy-spotify are using. Which goes back to my
earlier point about checking the install path. Perhaps you have installed
it multiple times, once from apt and once from git.

On Fri, 12 Oct 2018, 00:15 Maxwell, <notifications@github.com> wrote:

> I am afraid I don't understand what you are asking. However, that last
> commit should fix your other case now too.
>
> I installed your fix, but it doesn't work for me.
> Let me explain, what i try to do.
>
> I have installed a mopidy, mopidy-spotify, all dependencies and WEB ui
> "Iris". This all together works fine at all.
>
> What i want to do is starting spotify with MPC command line.
>
> sudo systemctl restart mopidy
> pi@box:~ $ sudo mpc -h box -p 6601 add "spotify:artist:0oLWXBuoO0namCRC2oji0q"
> pi@box:~ $ sudo mpc -h box -p 6601 clear
> volume:100%   repeat: off   random: off   single: off   consume: off
> pi@box:~ $ sudo mpc -h box -p 6601 add "spotify:playlist:37i9dQZF1DX0XUsuxWHRQd"
> error adding spotify:playlist:37i9dQZF1DX0XUsuxWHRQd: directory or file not found
> pi@box:~ $ sudo mpc -h box -p 6601 add "spotify:user:spotify:playlist:37i9dQZF1DX0XUsuxWHRQd"
> error adding spotify:user:spotify:playlist:37i9dQZF1DX0XUsuxWHRQd: directory or file not found
>
> you see, adding artist works fine. albums and tracks too.
> but when i try to add a playlist with MPC, i get an error.
>
> for own testing, i removed mopidy-spotify (not his dependencies) from my
> box. i restarted mopidy. the functionality of MPC was the same like before,
> everything ok but playlists don't work.
>
> so the question is, does MPC use mopidy-spotify für interpeting Spotify
> URIs?? or does MPC use pyspotify2.0.5 and libspotify12 for this???
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mopidy/mopidy-spotify/pull/188#issuecomment-429152828>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA5DqFTUiH3k7bgx8sLqbxqqUduhUv8jks5uj9EjgaJpZM4Uiou9>
> .
>

princemaxwell commented:
Damn, i checked the deps and found a second version of mopidy-spotify.
To remove all versions i first did "sudo apt-get remove mopidy-spotify" and then i manually removed "/usr/local/lib/python2.7/dist-packages/Mopidy_Spotify-3.1.0-py2.7.egg".

After that, mopidy told me, that the extension mopidy-spotify is missing.
Ok.

Then i started installation of the master branch. 

```
pi@box:~ $ sudo apt-get remove mopidy-spotify                               Reading package lists... Done
Building dependency tree
Reading state information... Done
Package 'mopidy-spotify' is not installed, so not removed
0 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
pi@box:~ $ sudo systemctl restart mopidy
pi@box:~ $ sudo apt-get install mopidy-spotify
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following NEW packages will be installed:
  mopidy-spotify
0 upgraded, 1 newly installed, 0 to remove and 2 not upgraded.
Need to get 0 B/23.6 kB of archives.
After this operation, 112 kB of additional disk space will be used.
WARNING: The following packages cannot be authenticated!
  mopidy-spotify
Install these packages without verification? [y/N] y
Selecting previously unselected package mopidy-spotify.
(Reading database ... 144958 files and directories currently installed.)
Preparing to unpack .../mopidy-spotify_3.1.0-0mopidy1_all.deb ...
Unpacking mopidy-spotify (3.1.0-0mopidy1) ...
Setting up mopidy-spotify (3.1.0-0mopidy1) ...
pi@box:~ $ sudo systemctl restart mopidy
```

At this point i manually overwrote the files in "/usr/lib/python2.7/dist-packages/mopidy_spotify/" with all the files from "kingosticks:fix/web_api_playlists/mopidy_spotify"...

```
pi@box:~ $ sudo systemctl restart mopidy

pi@box:~ $ mpc -h lianbox -p 6601 add spotify:user:spotify:playlist:37i9dQZF1DWUVpAXiEPK8P
error adding spotify:user:spotify:playlist:37i9dQZF1DWUVpAXiEPK8P: directory or file not found

pi@box:~ $ mpc -h lianbox -p 6601 clear
volume:100%   repeat: off   random: off   single: off   consume: off

pi@box:~ $ sudo mopidyctl deps
Running "/usr/bin/mopidy --config /usr/share/mopidy/conf.d:/etc/mopidy/mopidy.conf deps" as user mopidy
/usr/local/lib/python2.7/dist-packages/pylast/__init__.py:51: UserWarning: You are using pylast with Python 2. Pylast will soon be Python 3 only. More info: https://github.com/pylast/pylast/issues/265
  UserWarning,
Executable: /usr/bin/mopidy
Platform: Linux-4.14.71-v7+-armv7l-with-debian-9.4
Python: CPython 2.7.13 from /usr/lib/python2.7
Mopidy: 2.2.0 from /usr/lib/python2.7/dist-packages
Mopidy-Local-Images: 1.0.0 from /usr/local/lib/python2.7/dist-packages
  Mopidy>=1.1: 2.2.0 from /usr/lib/python2.7/dist-packages
  setuptools: 33.1.1 from /usr/lib/python2.7/dist-packages
  Pykka>=1.1: 1.2.1 from /usr/lib/python2.7/dist-packages
  uritools>=1.0: 2.2.0 from /usr/local/lib/python2.7/dist-packages
    ipaddress; python_version == "2.7": 1.0.17 from /usr/lib/python2.7/dist-packages
Mopidy-Spotify: 3.1.0 from /usr/lib/python2.7/dist-packages
  Mopidy>=2.0: 2.2.0 from /usr/lib/python2.7/dist-packages
  Pykka>=1.1: 1.2.1 from /usr/lib/python2.7/dist-packages
  pyspotify>=2.0.5: 2.0.5 from /usr/lib/python2.7/dist-packages
    cffi>=1.0.0: 1.9.1 from /usr/lib/python2.7/dist-packages
  requests>=2.0: 2.12.4 from /usr/lib/python2.7/dist-packages
Mopidy-Iris: 3.27.1 from /usr/local/lib/python2.7/dist-packages
  tornado<5.0,>=3.2: 4.4.3 from /usr/lib/python2.7/dist-packages
  setuptools>=3.3: 33.1.1 from /usr/lib/python2.7/dist-packages
  ConfigObj>=5.0.6: 5.0.6 from /usr/local/lib/python2.7/dist-packages
    six: 1.10.0 from /usr/lib/python2.7/dist-packages
  requests>=2.0.0: 2.12.4 from /usr/lib/python2.7/dist-packages
  Mopidy>=2.0: 2.2.0 from /usr/lib/python2.7/dist-packages
  Mopidy-Local-Images>=1.0: 1.0.0 from /usr/local/lib/python2.7/dist-packages
    Mopidy>=1.1: 2.2.0 from /usr/lib/python2.7/dist-packages
    setuptools: 33.1.1 from /usr/lib/python2.7/dist-packages
    Pykka>=1.1: 1.2.1 from /usr/lib/python2.7/dist-packages
    uritools>=1.0: 2.2.0 from /usr/local/lib/python2.7/dist-packages
      ipaddress; python_version == "2.7": 1.0.17 from /usr/lib/python2.7/dist-packages
  pylast>=1.6.0: 2.4.0 from /usr/local/lib/python2.7/dist-packages
    six: 1.10.0 from /usr/lib/python2.7/dist-packages
GStreamer: 1.10.4.0 from /usr/lib/python2.7/dist-packages/gi
  Detailed information:
    Python wrapper: python-gi 3.22.0
    Relevant elements:
      Found:
        uridecodebin
        souphttpsrc
        appsrc
        alsasink
        osssink
        oss4sink
        pulsesink
        id3demux
        id3v2mux
        lamemp3enc
        mad
        mpegaudioparse
        mpg123audiodec
        vorbisdec
        vorbisenc
        vorbisparse
        oggdemux
        oggmux
        oggparse
        flacdec
        flacparse
        shout2send
      Not found:
        flump3dec
```

What did i wrong?
Was my way the correct way to get your fix running?

If not, can you explain, me how to get your fix installed correctly?
MANY THANKS!

princemaxwell commented:
@kingosticks AHHHHHHH!
I got it! You will never believe, what i had to do...
I used an old client_id and client_secret from developer.spotify.com and i had to renew this on mopidy.com/authenticate/ and then playlists are working FINE.

Thanks a lot, this was my fault. But i never thought about about that...

princemaxwell commented:
@kingosticks 
ok, it works now, but the start of mopidy is very very slow.

`INFO     Starting Mopidy 2.2.0
INFO     Loading config from builtin defaults
INFO     Loading config from /home/pi/.config/mopidy/mopidy.conf
INFO     Loading config from command line options
INFO     Enabled extensions: iris, mpd, http, stream, spotify, m3u, softwaremixer, file, local-images, local
INFO     Disabled extensions: none
INFO     Starting Mopidy mixer: SoftwareMixer
INFO     Starting Mopidy audio
INFO     Starting Mopidy backends: StreamBackend, M3UBackend, FileBackend, LocalBackend, SpotifyBackend
INFO     Audio output set to "alsasink"
INFO     No local library metadata cache found at /home/pi/.local/share/mopidy/local/library.json.gz. Please run `mopidy local scan` to index your local music library. If you do not have a local music collection, you can disable the local backend to hide this message.
INFO     Loaded 0 local tracks using json
INFO     Logged in to Spotify in offline mode
INFO     Logged in to Spotify in online mode
INFO     Logged into Spotify Web API as spotifyuser
INFO     Loaded 47 playlists
INFO     Refresh Playlists took 14812ms
INFO     Starting Mopidy core
INFO     Starting Mopidy frontends: IrisFrontend, MpdFrontend, HttpFrontend
INFO     Starting Iris 3.27.1
INFO     MPD server running at [::ffff:0.0.0.0]:6601
INFO     Starting GLib mainloop
INFO     HTTP server running at [::ffff:0.0.0.0]:6680
`

Why does mopidy-spotify load my playlists, whats takes 15seconds.
Can i skip this step?

kingosticks commented:
It loads all your playlists yes. It could do this in the background like libspotify did.

As to why iris takes so long I'm not sure. Perhaps it's doing a lot of extra playlist lookups. But once that's all done it should be cached and it should be quick to browse, much more usable than deferring loading until you select the playlist. Typically mopidy is starting as you system starts so it it takes 15 seconds I don't think that's a big issue.

princemaxwell commented:
When i replace your branch with the master branch, mopidy is starting in 5 seconds and Iris is available after 10-15 seconds.
Why does it load all my playlists? My intention is to load a playlist via mpc...

kingosticks commented:
Yeh, and none of the playlists work so that's really not an interesting data point.

Loading all your playlists once at startup caches the data and makes selecting a playlist later much, much faster. If you only load then on demand then you'll have to wait a few seconds, much longer for big playlists, and its a much worse user experience.

I will do some testing with iris myself and see what it's doing differently to the other clients.

princemaxwell commented:
@kingosticks i have build a music box for the kids with RFID card reader to start music by swiping cards easily.

The box starts and a start sound appears. this should be the signal for the kids, that the box is ready. If configured the startsound.service to start after Mopidy.service. But the sound is played minutes before Mopidy is ready to use.

So the first problem is, that Mopidy needs much more time to start completely with this playlist_fix as without.

I use the MPD from Mopidy to play local MP3, livestreams and Spotify.

The second problem is, that the MP3 and livestream loading is slow with MPD + Mopidy.
When I only use MPD without Mopidy, MP3 and livestreams are loading fast.

I hope that I explained it correct now...

kingosticks commented:
OK thanks. In general I would suggest you implement a more robust solution that polls mopidy and waits for it to respond before playing the sound. With the latest code it will appear ready much earlier but ultimately any Spotify playlist actions won't be successful until the initial (now background) playlist loading completes.

As to the second issue I still don't understand but it sounds like a very separate issue so let's not discuss that here.

princemaxwell commented:
Feedback: I tested your last commit. Mopidy now starts in 8 seconds. That's great.
My problems with MP3 and livestreams are not a problem of mopidy-spotify, because they occur even if mopidy-spotify is not installed.
So, that is a problem of mopidy...

girst commented:
I don't think the async loading of playlists is that great. It makes it much harder to script the execution of a playlist (e.g. wait for mopidy to be started, then load and start a playlist). 

What good comes from a fast startup when it isn't usable afterwards?

kingosticks commented:
Ideally you will wait for the playlists loaded event to fire, *if* whatever you want to do involves playlists.

ikcalB commented:
any progress merging this into master?

L422Y commented:
Does this sync more than 100 tracks from a playlist? The current fix I have in place seems to be cut off at 100 tracks (this playlist should have 500+):

Still Woozy - Goodie Bag
[playing] #76/100   1:52/2:26 (76%)
volume:100%   repeat: on    random: on    single: off   consume: off
ubuntu@ubuntu:~$

BlackLight commented:
@L422Y I've fixed that on my fork - the Spotify API returns 100 tracks max per playlist and requires another request to get another chunk of 100 tracks. Probably a similar fix needs to be applied here as well.

@kingosticks I know I'm becoming repetitive, but is there any progress/timeline on merging this into master? Almost a year and a half has gone since the original report :)

kingosticks commented:
@BlackLight I do hear you. I've been using my branch with the latest Mopidy develop and have been running into a couple of issues. I think they are with Mopidy itself rather than my changes but I've been working to get them sorted first.

Could you point me to the fix for this 100 tracks issue?

BlackLight commented:
@kingosticks I remember that I struggled a bit back then to get pagination working. I eventually managed to get it working with https://github.com/BlackLight/mopidy-spotify/commit/286ea9ae0923b8fafd3ecd5ac6cce710f1d456b6 (for playlist expansion), https://github.com/BlackLight/mopidy-spotify/commit/748c019ae3b88d7e821ec471c8bf770de7d31cdb (to support the expansion of the `next` page) and https://github.com/BlackLight/mopidy-spotify/commit/8683e38d4b60df99daaa0736968751c691acdb64 (to make the `_lookup_playlist` method a bit more compact).

kingosticks commented:
Hmm OK, I thought that was all entirely handled already, but obviously it has been a while... I'll look again more closely.

kingosticks commented:
I don't have this playlist length issue. Perhaps you were using an old version.

L422Y commented:
I updated to the newer version after we did all the manual adjustments and it is working, i just have custom mods that push events to the OS to start playback, and updates to a Slack channel for when the track changes, so I was "nervous" about those, but all is well.
